### ISSUE TITLE
No error for "popd .."

### ISSUE THREAD

**Issue by memotype** [koalaman/shellcheck#1512]
#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

Should give an error when `popd` is given an invalid argument (only accepts `popd [-n] [+N | -N]`). This can happen when editing a script to change `cd`s to `pushd`/`popd`. I.e:

#### Here's a snippet or screenshot that shows the problem:

Editing:

```sh
#!/bin/bash
cd foo || exit 1
do something
cd ..
do something else
```

To:

```sh
#!/bin/bash
pushd foo || exit 1
do something
popd ..
do something else
```

Bash will give an error like this:

```
-bash: popd: ..: invalid argument
```

Can be an easy mistake to make that shellcheck could catch.

**Comment by koalaman** [comment]
You're right of course. I had a look at 20k random scripts, and this does not appear to be a systemic issue: I found three instances of popd used with arguments, and they were all variables so it's hard to autodetect whether they're actual true positives. Given the low rate, it's probably best left to `popd`'s fairly descriptive error message.

Thanks for the suggestion!

**Comment by memotype** [comment]
Ok, but, I thought the idea of shellcheck was to detect errors or possible issues before they were discovered at runtime? Feels like waiting until the runtime issues an error misses the point of having a linter-like system for shell scripts? Imagine a script which does a "popd ..", but without "set -e", and then does a "rm *". In a test environment, this might not be caught as an issue, and when run in production, wipes out important data. I thought I was making a fairly simple suggestion of something which shellcheck could possibly help prevent.

Would it be more likely to be included if I submitted a PR? I'm not entirely unfamiliar with Haskell and could give it a try if desired.

**Comment by koalaman** [comment]
It would be great if ShellCheck allowed you to formally verify that a script is correct or get close enough that you don't even have to test it. Unfortunately, I don't think that's possible, and it's therefore not one of ShellCheck's stated goals (listed in the README.md).

All checks come with an associated cost, in terms of code size, code complexity, runtime, and noise. Each tries to justify its existence in terms of how often it triggers and how helpful it is when it does.

If the mistake was seen frequently (1+ in 1000 scripts) then it would be an immediate shoo-in, but it's instead exceedingly rare. If it was an especially dangerous mistake, it might as well. However, in the scenario you describe, the user would already get a SC2164 suggesting you handle failures in `popd`, so any bad effects are already essentially mitigated.

For comparison, SC2114 (`rm -rf /usr /lib/foo.so.1`) also triggers at a <1 in 20,000 kind of frequency, but it prevents the loss of major system directories. SC2061 (`find -name *.txt`) also mostly just helps with what would have been a fairly straight forward error message (on GNU), but does so in 1 in 250 scripts.

The check is definitely more likely to be included if you submit a PR because I try my best to encourage contributions as an investment, but I really wish you'd choose an issue that will be more helpful to more people.

(My triaging could always be way off of course, so do let me know if you e.g. see several instances of this happening on GitHub or elsewhere)

**Comment by memotype** [comment]
Fair enough, and thank you for taking the time to explain the reasoning.
